src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,198 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\ConstantType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\NullType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use function strtolower;
+
+final class CurlGetinfoFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function __construct(
+		private ReflectionProvider $reflectionProvider,
+	)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return strtolower($functionReflection->getName()) === 'curl_getinfo';
+	}
+
+	/**
+	 * @throws ShouldNotHappenException
+	 */
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (count($functionCall->getArgs()) < 1) {
+			return ParametersAcceptorSelector::selectSingle(
+				$functionReflection->getVariants(),
+			)->getReturnType();
+		}
+
+		if (count($functionCall->getArgs()) > 1) {
+			$componentType = $scope->getType($functionCall->getArgs()[1]->value);
+
+			if (!$componentType instanceof ConstantType) {
+				return $this->createAllComponentsReturnType();
+			}
+
+			$componentType = $componentType->toInteger();
+			if (!$componentType instanceof ConstantIntegerType) {
+				throw new ShouldNotHappenException();
```

Comment:
Wrong calls with a "string" or similar should not throw an exception - that leads to internal error. Please test this in `tests/PHPStan/Analyser/data/curl_getinfo.php` and return the default return type instead.